File: packages/ts/src/rules/possession-rules.ts
import type { Issue, OcfDoc } from "../types.js";
import { getFrames, type DocContext } from "../context.js";
import type { FrameState } from "../possession.js";
import { makeIssue } from "../codes.js";

const BALL_DEPENDENT = new Set(["pass", "shoot", "dribble"]);
const PICKUP = new Set(["pickup", "rebound"]);

// Returns the resolved ball id, or the "AMBIGUOUS" sentinel (NOT a real ball id)
// when ball_id is omitted with multiple balls in play, or null when no ball exists.
function resolveBallId(action: Record<string, unknown>, ctx: DocContext): string | "AMBIGUOUS" | null {
  if (typeof action.ball_id === "string") return action.ball_id;
  if (ctx.ballIds.size === 1) return [...ctx.ballIds][0];
  if (ctx.ballIds.size === 0) return null;
  return "AMBIGUOUS";
}

export function possessionRules(doc: OcfDoc, ctx: DocContext, states: FrameState[]): Issue[] {
  const issues: Issue[] = [];
  const frames = getFrames(doc);

  frames.forEach((frame, fi) => {
    const frameId = frame.id as string | undefined;
    const state = states[fi];
    const actions = (frame.actions ?? []) as Record<string, unknown>[];

    // Intra-frame possession: actions within a frame execute in sequence, so an
    // earlier transfer (e.g. a `pass` to a teammate who then `shoot`s on the
    // catch) changes who currently carries the ball BEFORE a later action is
    // checked. We seed a mutable carrier map from the frame-start snapshot and
    // advance it as we walk the actions, so "currently carries" (design §B) is
    // evaluated at the point each action runs — not only at frame entry. The
    // frame's own end_state is still the cross-frame source of truth (engine
    // snapshots), so we never write back here.
    const carrier = new Map<string, string | null>();
    const loose = new Set<string>();
    for (const id of ctx.ballIds) {
      carrier.set(id, state.carrierOf(id));
      if (state.looseAt(id) !== null) loose.add(id);
    }

    actions.forEach((action, ai) => {
      const type = action.type as string;
      const player = action.player as string;
      const path = `/frames/${fi}/actions/${ai}`;

      if (BALL_DEPENDENT.has(type)) {
        const ball = resolveBallId(action, ctx);
        if (ball === "AMBIGUOUS") {
          issues.push(makeIssue("BALL_AMBIGUOUS", path, { player, count: ctx.ballIds.size }, frameId));
          return;
        }
        if (ball && carrier.get(ball) !== player) {
          issues.push(makeIssue("BALL_CARRIER_MISMATCH", path,
            { player, action: type, ball_id: ball }, frameId));
        }
        if (ctx.entityRefs.get(player)?.type === "defense") {
          issues.push(makeIssue("ACTION_UNUSUAL_CARRIER", path, { player, action: type }, frameId));
        }
        // Apply the action's effect on possession for subsequent actions.
        if (ball) applyEffect(type, player, action, ball, carrier, loose);
      } else if (PICKUP.has(type)) {
        const ball = resolveBallId(action, ctx);
        if (ball === "AMBIGUOUS") {
          issues.push(makeIssue("BALL_AMBIGUOUS", path, { player, count: ctx.ballIds.size }, frameId));
        } else if (ball && !loose.has(ball)) {
          // pickup/rebound needs a LOOSE ball. If it's not loose (carried by someone,
          // or dead) it cannot be picked up. (An unknown ball id is separately caught
          // by REF_BALL_UNKNOWN, so no guard is needed here.)
          issues.push(makeIssue("BALL_NOT_AT_LOCATION", path, { player, action: type }, frameId));
        } else if (ball) {
          applyEffect(type, player, action, ball, carrier, loose);
        }
      }
    });
  });
  return issues;
}

// Advance the intra-frame possession model after an action has been checked.
function applyEffect(
  type: string,
  player: string,
  action: Record<string, unknown>,
  ball: string,
  carrier: Map<string, string | null>,
  loose: Set<string>,
): void {
  switch (type) {
    case "pass": {
      const to = action.to_player;
      // After a pass the receiver carries the ball (in flight is modelled as the
      // receiver's possession for the purpose of a catch-and-shoot in the frame).
      carrier.set(ball, typeof to === "string" ? to : null);
      loose.delete(ball);
      break;
    }
    case "shoot":
      // The ball leaves the shooter; it is no longer carried.
      carrier.set(ball, null);
      loose.delete(ball);
      break;
    case "pickup":
    case "rebound":
      carrier.set(ball, player);
      loose.delete(ball);
      break;
    // `dribble` keeps the same carrier — no change.
  }
}

File: packages/py/ocf_validator/v2/possession_rules.py
from typing import Any

from ..codes import make_issue
from ..types import Issue
from .context import DocContextV2, is_branch

_BALL_DEPENDENT = {"pass", "shoot", "dribble"}
_PICKUP = {"pickup", "rebound"}
_NO_BALL_MOVEMENT = {"move", "cut"}

# screen/defend/tackle/faceoff/check are intentionally untracked here: none
# carries a ball_id in the schema, so there is no possession effect to apply
# or check for them (this differs from _NO_BALL_MOVEMENT, which actively
# checks that the actor is NOT holding a ball; these five get no check at
# all, since a screen/defend/etc. isn't disqualified by holding a ball).
# action_clear ALSO has a ball_id (an invasion-sport ball-clearing action)
# and would otherwise fit _BALL_DEPENDENT's shape, but is left untracked
# deliberately: the schema itself $comments it "Reserved for invasion sports
# ... No variants defined yet" (schema/v1.json, action_clear) -- i.e. it's a
# placeholder type with no finalized semantics yet, not a basketball action
# this validator currently needs to support. Revisit once invasion-sport
# action types are promoted out of "reserved" status.

_AMBIGUOUS = "AMBIGUOUS"


def _resolve_ball_ids(action: dict[str, Any], ctx: DocContextV2) -> list[str] | str:
    """Resolved ball ids, or the sentinel "AMBIGUOUS" (not real ball ids) when
    neither ball_id nor ball_ids is given and multiple balls exist in the
    document. Mirrors v1's _resolve_ball_id (single ball in doc -> assume it;
    zero balls -> no check; 2+ balls with nothing given -> AMBIGUOUS),
    extended for v2's ball_ids (plural, two-ball dribble) form."""
    bid = action.get("ball_id")
    if isinstance(bid, str):
        return [bid]
    bids = action.get("ball_ids")
    if isinstance(bids, list):
        ids = [b for b in bids if isinstance(b, str)]
        if ids:
            return ids
    if len(ctx.ball_ids) == 1:
        return list(ctx.ball_ids)
    if len(ctx.ball_ids) == 0:
        return []
    return _AMBIGUOUS


def _player_holds_any_ball(player: str, carrier: dict[str, str | None]) -> bool:
    return any(holder == player for holder in carrier.values())


def _apply_effect(
    type_: str,
    player: str,
    action: dict[str, Any],
    ball_ids: list[str],
    carrier: dict[str, str | None],
    loose: set[str],
) -> None:
    for ball in ball_ids:
        if type_ == "pass":
            to = action.get("to_player")
            carrier[ball] = to if isinstance(to, str) else None
            loose.discard(ball)
        elif type_ == "shoot":
            carrier[ball] = None
            loose.discard(ball)
        elif type_ in ("pickup", "rebound"):
            carrier[ball] = player
            loose.discard(ball)
        # `dribble` keeps the same carrier -- no change.


def possession_rules_v2(doc: dict[str, Any], ctx: DocContextV2) -> list[Issue]:
    issues: list[Issue] = []

    carrier: dict[str, str | None] = {}
    loose: set[str] = set()
    for b in doc.get("balls") or []:
        if not isinstance(b, dict):
            continue
        bid = b.get("id")
        if not isinstance(bid, str):
            continue
        carried_by = b.get("carried_by")
        carrier[bid] = carried_by if isinstance(carried_by, str) else None
        if b.get("at") is not None:
            loose.add(bid)

    def _check_and_apply(
        item: dict[str, Any],
        path: str,
        local_carrier: dict[str, str | None],
        local_loose: set[str],
    ) -> None:
        if is_branch(item):
            cases = item.get("cases") or {}
            for outcome, branch_case in cases.items():
                # Fork state per case: each outcome is a hypothetical alternate
                # continuation, not a sequential extension shared with sibling
                # cases. A case's fork is also a dead end for possession
                # purposes: we never merge it back into the top-level walk
                # after the branch returns, so the top-level sequence
                # continues from its own pre-branch state, oblivious to what
                # any case did. `then` (resuming the flat sequence at an
                # earlier/later action id) is a control-flow construct
                # handled elsewhere, not a possession-state merge point.
                forked_carrier = dict(local_carrier)
                forked_loose = set(local_loose)
                nested = branch_case.get("actions") or []
                for i, child in enumerate(nested):
                    _check_and_apply(child, f"{path}/cases/{outcome}/actions/{i}", forked_carrier, forked_loose)
            return

        type_ = item.get("type")
        player = item.get("player")

        if type_ in _NO_BALL_MOVEMENT and _player_holds_any_ball(player, local_carrier):
            issues.append(make_issue("BALL_CARRIER_MISMATCH", path, {"player": player, "action": type_}))
            return

        if type_ in _BALL_DEPENDENT:
            ball_ids = _resolve_ball_ids(item, ctx)
            if ball_ids == _AMBIGUOUS:
                issues.append(make_issue("BALL_AMBIGUOUS", path, {"player": player, "count": len(ctx.ball_ids)}))
                return
            for ball in ball_ids:
                if local_carrier.get(ball) != player:
                    issues.append(
                        make_issue("BALL_CARRIER_MISMATCH", path, {"player": player, "action": type_, "ball_id": ball})
                    )
            info = ctx.entity_refs.get(player)
            if info and info.type == "defense":
                issues.append(make_issue("ACTION_UNUSUAL_CARRIER", path, {"player": player, "action": type_}))
            _apply_effect(type_, player, item, ball_ids, local_carrier, local_loose)
        elif type_ in _PICKUP:
            ball_ids = _resolve_ball_ids(item, ctx)
            if ball_ids == _AMBIGUOUS:
                issues.append(make_issue("BALL_AMBIGUOUS", path, {"player": player, "count": len(ctx.ball_ids)}))
                return
            for ball in ball_ids:
                if ball not in local_loose:
                    issues.append(make_issue("BALL_NOT_AT_LOCATION", path, {"player": player, "action": type_}))
                else:
                    _apply_effect(type_, player, item, [ball], local_carrier, local_loose)

    top_level = doc.get("actions") or []
    for i, item in enumerate(top_level):
        _check_and_apply(item, f"/actions/{i}", carrier, loose)

    return issues
